utils/babeltrace_thapi.in

```diff
@@ -1,118 +1,194 @@
-
-$event_lambdas = {}
-require 'babeltrace_opencl_lib'
-require 'babeltrace_ze_lib'
-require 'babeltrace_cuda_lib'
-require 'babeltrace_omp_lib'
-
-$options = {
-  context: false,
-  restrict: false,
-  live: false
-}
-
-OptionParser.new do |opts|
-  opts.banner = "Usage: babeltrace_thapi [OPTIONS] target_trace_directory"
 
-  opts.on("-c", "--[no-]context", "Add context information") do |context|
-    $options[:context] = context
-  end
+# Don't complain about broken pipe
+Signal.trap('SIGPIPE', 'SYSTEM_DEFAULT')
+
+subtext = <<~HELP
+  Commonly used command are:
+     trace:     Display the trace
+     tally:     Display the tally
+     timeline:  Dump the timeline
+
+  See 'babeltrace_thapi COMMAND --help' for more information on a specific command.
+HELP
+
+$options = { backends: %w[omp cl ze cuda] }
+$options_tally = {
+  'display' => [String, 'compact'],
+  'name' => [String, 'demangle'],
+  'display_mode' => [String, 'human'],
+  'display_metadata' => [FalseClass, false],
+  'display_name_max_size' => [Integer, 80],
+  'display_kernel_verbose' => [FalseClass, false]
+}
 
-  opts.on("-r", "--[no-]restrict", "Restrict output to recognized events") do |restrict|
-    $options[:restrict] = restrict
+def common_options(opts)
+  opts.on('-b', '--backend BACKEND', Array) do |v|
+    $options[:backends] = v
   end
 
-  opts.on("-h", "--help", "Prints this help") do
-    puts opts
+  opts.on('-v', '--version', 'Print the version string') do
+    puts File.read(File.join(DATADIR, 'version'))
     exit
   end
+end
 
-  opts.on("--live", "Enable live display of the trace") do
-    $options[:live] = true
-  end
+subcommands = {
+  'trace' => OptionParser.new do |opts|
+               opts.banner = 'Usage: babeltrace_thapi [OPTIONS] trace_directory...'
+
+               opts.on('-c', '--[no-]context', 'Add context information') do |context|
+                 $options[:context] = context
+               end
+
+               opts.on('-r', '--[no-]restrict', 'Restrict output to recognized events') do |restrict|
+                 $options[:restrict] = restrict
+               end
+
+               opts.on('-h', '--help', 'Prints this help') do
+                 puts opts
+                 exit
+               end
+
+               opts.on('--live', 'Enable live display of the trace') do
+                 $options[:live] = true
+               end
+
+               common_options(opts)
+             end,
+  'tally' => OptionParser.new do |opts|
+               opts.banner = 'Usage: tally [OPTIONS] trace_directory...'
+
+               $options_tally.each do |k, (t, _)|
+                 opts.on("--#{k}=VALUE", t) do |v|
+                   $options_tally[k] = [t, v]
+                 end
+               end
+               common_options(opts)
+             end,
+  'timeline' => OptionParser.new do |opts|
+                  opts.banner = 'Usage: timeline [OPTIONS] trace_directory...'
+                  common_options(opts)
+                end
```

Comment:
Fix indentation here.

Reply:
how much intend? they are inside a dict, so 2 or 4 or 6?

Reply:
I would use the same euristic as for the =:
```ruby
  'trace' =>
    OptionParser.new do |opts|
    end
  'tally' =>
```

Reply:
Good idea!